helix-core/src/search.rs

```diff
@@ -61,3 +61,184 @@ pub fn find_nth_prev(
 
     Some(pos)
 }
+
+use crate::movement::Direction;
+use regex_automata::{dense, DenseDFA, Error as RegexError, DFA};
+use std::ops::Range;
+
+#[derive(Debug)]
+pub struct Searcher {
+    /// Locate end of match searching right.
+    right_fdfa: DenseDFA<Vec<usize>, usize>,
+    /// Locate start of match searching right.
+    right_rdfa: DenseDFA<Vec<usize>, usize>,
+
+    /// Locate start of match searching left.
+    left_fdfa: DenseDFA<Vec<usize>, usize>,
+    /// Locate end of match searching left.
+    left_rdfa: DenseDFA<Vec<usize>, usize>,
+}
+
+impl Searcher {
+    pub fn new(pattern: &str) -> Result<Searcher, RegexError> {
+        // Check case info for smart case
+        let has_uppercase = pattern.chars().any(|c| c.is_uppercase());
+
+        // Create Regex DFAs for all search directions.
+        let mut builder = dense::Builder::new();
+        let builder = builder.case_insensitive(!has_uppercase);
+
+        let left_fdfa = builder.clone().reverse(true).build(pattern)?;
+        let left_rdfa = builder
+            .clone()
+            .anchored(true)
+            .longest_match(true)
+            .build(pattern)?;
+
+        let right_fdfa = builder.clone().build(pattern)?;
+        let right_rdfa = builder
+            .anchored(true)
+            .longest_match(true)
+            .reverse(true)
+            .build(pattern)?;
+
+        Ok(Searcher {
+            right_fdfa,
+            right_rdfa,
+            left_fdfa,
+            left_rdfa,
+        })
+    }
+    pub fn search_prev(&self, text: RopeSlice, offset: usize) -> Option<Range<usize>> {
+        let text = text.slice(..offset);
+        let start = self.rfind(text, &self.left_fdfa)?;
+        let end = self.find(text.slice(start..), &self.left_rdfa)?;
+
+        Some(start..start + end)
+    }
+
+    pub fn search_next(&self, text: RopeSlice, offset: usize) -> Option<Range<usize>> {
+        let text = text.slice(offset..);
+        let end = self.find(text, &self.right_fdfa)?;
+        let start = self.rfind(text.slice(..end), &self.right_rdfa)?;
+
+        Some(offset + start..offset + end)
+    }
+
+    /// Returns the end offset of the longest match. If no match exists, then None is returned.
+    /// NOTE: based on DFA::find_at
+    fn find(&self, text: RopeSlice, dfa: &impl DFA) -> Option<usize> {
```

Comment:
I believe this code is running afoul of this warning here: https://docs.rs/regex-automata/0.1.10/regex_automata/dense/enum.DenseDFA.html#the-dfa-trait

I would suggest unwrapping the `DenseDFA` enum into a [`PremultipliedByteClass`](https://docs.rs/regex-automata/0.1.10/regex_automata/dense/struct.PremultipliedByteClass.html), and then just that everywhere.

(In `regex-automata 0.2`, this perf footgun will disappear since a dense DFA will no longer be an enum. It will just be a `PremultipliedByteClass`.)

Reply:
Good point, I should have paid more attention to the docs.

What else are you changing in `0.2`?

Reply:
It's a total rewrite. This is the high level plan: https://github.com/rust-lang/regex/issues/656

For DFAs in particular, they are getting support for everything except for Unicode word boundary. (So `(?-u:\b)`, `^` and `$` will work.) And they will also have multi-pattern support. But otherwise, all DFAs will be pre-multiplied with byte classes enabled, and state IDs will always be `u32`. (So there are much fewer type parameters.)

And lots of compilation perf improvements. But not enough to make building large Unicode patterns bearable.